### ISSUE TITLE
build error: cl-acclimation/packages.lisp which is not relative to code root in system CLASP-CLEAVIR

### ISSUE THREAD

**Issue by Ambrevar** [clasp-developers/clasp#1366]
**Describe the bug**

I'm in the process of packaging Clasp for Guix.
When running

```sh
./koga --skip-sync
```

(which is required because Guix builds in offline containers) I get

```
While evaluating the form starting at line 193, column 0
  of #P"/tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/src/lisp/cscript.lisp":
Unhandled SIMPLE-ERROR in thread #<SB-THREAD:THREAD "main thread" RUNNING
                                    {1002468113}>:
  Found source path of /gnu/store/50ckbwgrr2sqimmjdwdwlpv0ln03k5kv-cl-acclimation-0.0.0-1.4d51150/share/common-lisp/source/cl-acclimation/packages.lisp which is not relative to code root in system CLASP-CLEAVIR.

Backtrace for: #<SB-THREAD:THREAD "main thread" RUNNING {1002468113}>
0: (SB-DEBUG::DEBUGGER-DISABLED-HOOK #<SIMPLE-ERROR "Found source path of ~a which is not relative to code root in system ~a." {10046950F3}> #<unused argument> :QUIT T)
1: (SB-DEBUG::RUN-HOOK *INVOKE-DEBUGGER-HOOK* #<SIMPLE-ERROR "Found source path of ~a which is not relative to code root in system ~a." {10046950F3}>)
2: (INVOKE-DEBUGGER #<SIMPLE-ERROR "Found source path of ~a which is not relative to code root in system ~a." {10046950F3}>)
3: (ERROR "Found source path of ~a which is not relative to code root in system ~a." #P"/gnu/store/50ckbwgrr2sqimmjdwdwlpv0ln03k5kv-cl-acclimation-0.0.0-1.4d51150/share/common-lisp/source/cl-acclimation/packages.lisp" :CLASP-CLEAVIR)
4: ((:METHOD KOGA::ADD-TARGET-SOURCE (T T SYMBOL)) #<KOGA::CONFIGURATION {100BA65153}> :CCLASP :CLASP-CLEAVIR) [fast-method]
5: (KOGA:SOURCES :CCLASP :CLASP-CLEAVIR src/lisp/kernel/cleavir/auto-compile.lisp src/lisp/kernel/cleavir/inline.lisp src/lisp/kernel/lsp/queue.lisp src/lisp/kernel/cmp/compile-file-parallel.lisp src/lisp/kernel/lsp/generated-encodings.lisp src/lisp/kernel/lsp/process.lisp src/lisp/kernel/lsp/top.lisp src/lisp/kernel/lsp/encodings.lisp src/lisp/kernel/lsp/cltl2.lisp)
6: (ADD-CCLASP-SOURCES :CCLASP)
7: (SIMPLE-EVAL-IN-LEXENV (ADD-CCLASP-SOURCES) #<NULL-LEXENV>)
8: (EVAL-TLF (ADD-CCLASP-SOURCES) 6 NIL)
9: ((LABELS SB-FASL::EVAL-FORM :IN LOAD-AS-SOURCE) (ADD-CCLASP-SOURCES) 6)
10: ((LAMBDA (FORM &KEY :CURRENT-INDEX &ALLOW-OTHER-KEYS) :IN LOAD-AS-SOURCE) (ADD-CCLASP-SOURCES) :CURRENT-INDEX 6)
11: (SB-C::%DO-FORMS-FROM-INFO #<FUNCTION (LAMBDA (FORM &KEY :CURRENT-INDEX &ALLOW-OTHER-KEYS) :IN LOAD-AS-SOURCE) {1003B8EA1B}> #<SB-C::SOURCE-INFO {1003B8E9E3}> SB-C::INPUT-ERROR-IN-LOAD)
12: (LOAD-AS-SOURCE #<FORM-TRACKING-STREAM for "file /tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/src/lisp/cscript.lisp" {1003B8D033}> :VERBOSE NIL :PRINT NIL :CONTEXT "loading")
13: ((LABELS SB-FASL::LOAD-STREAM-1 :IN LOAD) #<FORM-TRACKING-STREAM for "file /tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/src/lisp/cscript.lisp" {1003B8D033}> NIL)
14: (SB-FASL::CALL-WITH-LOAD-BINDINGS #<FUNCTION (LABELS SB-FASL::LOAD-STREAM-1 :IN LOAD) {7FFFF7436D5B}> #<FORM-TRACKING-STREAM for "file /tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/src/lisp/cscript.lisp" {1003B8D033}> NIL #<FORM-TRACKING-STREAM for "file /tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/src/lisp/cscript.lisp" {1003B8D033}>)
15: (LOAD #P"src/lisp/cscript.lisp" :VERBOSE NIL :PRINT NIL :IF-DOES-NOT-EXIST :ERROR :EXTERNAL-FORMAT :DEFAULT)
16: (KOGA::DO-SETUP)
17: (KOGA::SETUP :SKIP-SYNC T :VERSION "1.0.0")
18: ((LAMBDA NIL :IN "/tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/koga"))
19: (SIMPLE-EVAL-IN-LEXENV (LET* ((INITARGS (NCONC (PARSE-COMMAND-LINE-ARGUMENTS) (IGNORE-ERRORS #) (IGNORE-ERRORS #))) (*GIT* (GETF INITARGS :GIT "git")) (BUILD (UIOP/PATHNAME:ENSURE-DIRECTORY-PATHNAME (GETF INITARGS :BUILD-PATH "build/"))) (EXTENSIONS (GETF INITARGS :EXTENSIONS)) (SKIP-SYNC (GETF INITARGS :SKIP-SYNC)) (HELP (GETF INITARGS :HELP))) (UNLESS HELP (FORMAT T "Synchronizing external repositories~%~%")) (LOOP FOR SOURCE IN (UIOP/STREAM:READ-FILE-FORM #P"repos.sexp") FOR NAME = (GETF SOURCE :NAME) FOR EXTENSION = ...) (WHEN (AND (NOT HELP) (GETF INITARGS :CLEAN)) (FORMAT T "Cleaning up previous build~%~%") (UIOP/FILESYSTEM:DELETE-DIRECTORY-TREE BUILD :VALIDATE T :IF-DOES-NOT-EXIST :IGNORE)) (ASDF/SOURCE-REGISTRY:INITIALIZE-SOURCE-REGISTRY (QUASIQUOTE (:SOURCE-REGISTRY (:TREE #) :INHERIT-CONFIGURATION))) (ASDF/OUTPUT-TRANSLATIONS:INITIALIZE-OUTPUT-TRANSLATIONS (QUASIQUOTE (:OUTPUT-TRANSLATIONS (T #) :INHERIT-CONFIGURATION))) (ASDF/OPERATE:LOAD-SYSTEM :KOGA) (APPLY (FUNCTION UIOP/PACKAGE:SYMBOL-CALL) "KOGA" (IF HELP "HELP" "SETUP") INITARGS)) #<NULL-LEXENV>)
20: (EVAL-TLF (LET* ((INITARGS (NCONC (PARSE-COMMAND-LINE-ARGUMENTS) (IGNORE-ERRORS #) (IGNORE-ERRORS #))) (*GIT* (GETF INITARGS :GIT "git")) (BUILD (UIOP/PATHNAME:ENSURE-DIRECTORY-PATHNAME (GETF INITARGS :BUILD-PATH "build/"))) (EXTENSIONS (GETF INITARGS :EXTENSIONS)) (SKIP-SYNC (GETF INITARGS :SKIP-SYNC)) (HELP (GETF INITARGS :HELP))) (UNLESS HELP (FORMAT T "Synchronizing external repositories~%~%")) (LOOP FOR SOURCE IN (UIOP/STREAM:READ-FILE-FORM #P"repos.sexp") FOR NAME = (GETF SOURCE :NAME) FOR EXTENSION = ...) (WHEN (AND (NOT HELP) (GETF INITARGS :CLEAN)) (FORMAT T "Cleaning up previous build~%~%") (UIOP/FILESYSTEM:DELETE-DIRECTORY-TREE BUILD :VALIDATE T :IF-DOES-NOT-EXIST :IGNORE)) (ASDF/SOURCE-REGISTRY:INITIALIZE-SOURCE-REGISTRY (QUASIQUOTE (:SOURCE-REGISTRY (:TREE #) :INHERIT-CONFIGURATION))) (ASDF/OUTPUT-TRANSLATIONS:INITIALIZE-OUTPUT-TRANSLATIONS (QUASIQUOTE (:OUTPUT-TRANSLATIONS (T #) :INHERIT-CONFIGURATION))) (ASDF/OPERATE:LOAD-SYSTEM :KOGA) (APPLY (FUNCTION UIOP/PACKAGE:SYMBOL-CALL) "KOGA" (IF HELP "HELP" "SETUP") INITARGS)) 10 NIL)
21: ((LABELS SB-FASL::EVAL-FORM :IN LOAD-AS-SOURCE) (LET* ((INITARGS (NCONC (PARSE-COMMAND-LINE-ARGUMENTS) (IGNORE-ERRORS #) (IGNORE-ERRORS #))) (*GIT* (GETF INITARGS :GIT "git")) (BUILD (UIOP/PATHNAME:ENSURE-DIRECTORY-PATHNAME (GETF INITARGS :BUILD-PATH "build/"))) (EXTENSIONS (GETF INITARGS :EXTENSIONS)) (SKIP-SYNC (GETF INITARGS :SKIP-SYNC)) (HELP (GETF INITARGS :HELP))) (UNLESS HELP (FORMAT T "Synchronizing external repositories~%~%")) (LOOP FOR SOURCE IN (UIOP/STREAM:READ-FILE-FORM #P"repos.sexp") FOR NAME = (GETF SOURCE :NAME) FOR EXTENSION = ...) (WHEN (AND (NOT HELP) (GETF INITARGS :CLEAN)) (FORMAT T "Cleaning up previous build~%~%") (UIOP/FILESYSTEM:DELETE-DIRECTORY-TREE BUILD :VALIDATE T :IF-DOES-NOT-EXIST :IGNORE)) (ASDF/SOURCE-REGISTRY:INITIALIZE-SOURCE-REGISTRY (QUASIQUOTE (:SOURCE-REGISTRY (:TREE #) :INHERIT-CONFIGURATION))) (ASDF/OUTPUT-TRANSLATIONS:INITIALIZE-OUTPUT-TRANSLATIONS (QUASIQUOTE (:OUTPUT-TRANSLATIONS (T #) :INHERIT-CONFIGURATION))) (ASDF/OPERATE:LOAD-SYSTEM :KOGA) (APPLY (FUNCTION UIOP/PACKAGE:SYMBOL-CALL) "KOGA" (IF HELP "HELP" "SETUP") INITARGS)) 10)
22: ((LAMBDA (FORM &KEY :CURRENT-INDEX &ALLOW-OTHER-KEYS) :IN LOAD-AS-SOURCE) (LET* ((INITARGS (NCONC (PARSE-COMMAND-LINE-ARGUMENTS) (IGNORE-ERRORS #) (IGNORE-ERRORS #))) (*GIT* (GETF INITARGS :GIT "git")) (BUILD (UIOP/PATHNAME:ENSURE-DIRECTORY-PATHNAME (GETF INITARGS :BUILD-PATH "build/"))) (EXTENSIONS (GETF INITARGS :EXTENSIONS)) (SKIP-SYNC (GETF INITARGS :SKIP-SYNC)) (HELP (GETF INITARGS :HELP))) (UNLESS HELP (FORMAT T "Synchronizing external repositories~%~%")) (LOOP FOR SOURCE IN (UIOP/STREAM:READ-FILE-FORM #P"repos.sexp") FOR NAME = (GETF SOURCE :NAME) FOR EXTENSION = ...) (WHEN (AND (NOT HELP) (GETF INITARGS :CLEAN)) (FORMAT T "Cleaning up previous build~%~%") (UIOP/FILESYSTEM:DELETE-DIRECTORY-TREE BUILD :VALIDATE T :IF-DOES-NOT-EXIST :IGNORE)) (ASDF/SOURCE-REGISTRY:INITIALIZE-SOURCE-REGISTRY (QUASIQUOTE (:SOURCE-REGISTRY (:TREE #) :INHERIT-CONFIGURATION))) (ASDF/OUTPUT-TRANSLATIONS:INITIALIZE-OUTPUT-TRANSLATIONS (QUASIQUOTE (:OUTPUT-TRANSLATIONS (T #) :INHERIT-CONFIGURATION))) (ASDF/OPERATE:LOAD-SYSTEM :KOGA) (APPLY (FUNCTION UIOP/PACKAGE:SYMBOL-CALL) "KOGA" (IF HELP "HELP" "SETUP") INITARGS)) :CURRENT-INDEX 10)
23: (SB-C::%DO-FORMS-FROM-INFO #<FUNCTION (LAMBDA (FORM &KEY :CURRENT-INDEX &ALLOW-OTHER-KEYS) :IN LOAD-AS-SOURCE) {100196E47B}> #<SB-C::SOURCE-INFO {100196E443}> SB-C::INPUT-ERROR-IN-LOAD)
24: (LOAD-AS-SOURCE #<FD-STREAM for "file /tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/./koga" {1001968CF3}> :VERBOSE NIL :PRINT NIL :CONTEXT "loading")
25: ((LABELS SB-FASL::LOAD-STREAM-1 :IN LOAD) #<FD-STREAM for "file /tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/./koga" {1001968CF3}> NIL)
26: (SB-FASL::CALL-WITH-LOAD-BINDINGS #<FUNCTION (LABELS SB-FASL::LOAD-STREAM-1 :IN LOAD) {7FFFF743780B}> #<FD-STREAM for "file /tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/./koga" {1001968CF3}> NIL #<FD-STREAM for "file /tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/./koga" {1001968CF3}>)
27: (LOAD #<FD-STREAM for "file /tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/./koga" {1001968CF3}> :VERBOSE NIL :PRINT NIL :IF-DOES-NOT-EXIST :ERROR :EXTERNAL-FORMAT :DEFAULT)
28: ((FLET LOAD-SCRIPT :IN PROCESS-SCRIPT) #<FD-STREAM for "file /tmp/guix-build-clasp-cl-1.0.0-1.f3b9992.drv-0/source/./koga" {1001968CF3}>)
29: ((FLET SB-UNIX::BODY :IN PROCESS-SCRIPT))
30: ((FLET "WITHOUT-INTERRUPTS-BODY-11" :IN PROCESS-SCRIPT))
31: (PROCESS-SCRIPT "./koga")
32: (TOPLEVEL-INIT)
33: ((FLET SB-UNIX::BODY :IN START-LISP))
34: ((FLET "WITHOUT-INTERRUPTS-BODY-3" :IN START-LISP))
35: (%START-LISP)
```

**Expected behavior**

If the system is available then I suppose it should not matter if it's a subdir or not.  Is there a good reason to keep this check?

At the very least I propose to disable it when `skip-sync` is on.

**Other steps to reproduce**

If the above does not make sense I can share the Guix build files to reproduce.

**Context**
* Versions or commit hashes of Clasp on which you observed the problem: f3b9992a5532e9bac6069212e4538ae17ce7132f
* If relevant, your operating system and other aspects of your computing environment: Guix

**Comment by yitzchak** [comment]
The build system is not designed to include the various cleavir dependencies from outside the build directory. Those dependencies are put into the installed Clasp source tree and marked as immutable ASDF systems.

Also, the build system depends on specific commits of those dependencies and sometimes git clones with slight changes.

Once the build is complete those dependencies are no longer needed and are now included in the startup image (called a "core" on sbcl) of clasp.

In short, you can't currently build with `--skip-sync` unless you already have all of the dependencies in the right places.

**Comment by Ambrevar** [comment]
This approach is sadly not very friendly with the various packaging systems out there that require build offline.

What's your recommendation?  Manually copy the deps at the right location?
Is there a way to get the exact list of dependencies, their versions and their expected location?

**Comment by yitzchak** [comment]
The git repos that are synced are in `repos.sexp`.  You can skip the ones labeled as extensions. The repos put into `src/` are probably the minimum list that are required to be in specific locations.
